### PR TITLE
Signal semaphore

### PR DESCRIPTION
**Pull request by Jorge-Rodriguez**
The semaphore mechanism we discussed yesterday.

### CODE REVIEW

File: see/observer.py [hunk 1]
```diff
@@ -47,8 +47,47 @@ def context(self):
 
 
 class Observable(Observatory):
+    class HashedSemaphore(object):
```
[the remark this answers is not in the document]
Comment (Jorge-Rodriguez) [reply see/observer.py]: From a performance point of view there is no difference in python. But semantically it conveys the idea that the nested class is to be used only in the context of its parent class.

**Comment by Jorge-Rodriguez** [comment]
There is an example on the tests, but the idea is that now we can use the `@handler` decorator on any given hook handler and that would ensure that the function acquires the semaphore for its event before executing and releases it afterwards.

But you are right that this is not thread safe, the semaphore operations should be atomic.

**Comment by noxdafox** [comment]
`@handler` seems a pretty generic term for a decorator which does much more.

What about `@semaphore`? Or `@refcount_handler`?

File: see/observer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (noxdafox) [reply see/observer.py]: I imagined that being the idea. This is a personal opinion so do not consider it a blocker for the PR.

To seclude classes within given scopes Python provides modules which can be used as namespaces. Nesting classes within each other IMHO makes them hardly readable with negligible benefits. The fact a class should be used internally within another one can be simply made explicit via documentation. On the downside, it becomes difficult to test the two classes separately, it makes them hardly readable (the parent class becomes big, the nested one suffers from the excessive indentation) and it couple the two entities making them less re-usable.

**Comment by Jorge-Rodriguez** [comment]
The idea behind the name `@handler` is that, as it decorates a hook's function, it indicates to the human reader "this is a handler function" while managing the semaphore behind the scenes.
I suppose it's a matter of what we want to communicate with the decorator name

File: see/observer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jorge-Rodriguez) [reply see/observer.py]: While I do see the point in your counter argument, I do not think the downsides apply to the current case. The Observable class hasn't become unmanageably big, The extra indentation does not make this particular inner class unreadable (line length is still concise) and while the two classes are indeed coupled I don't see an issue with that in this case, as the HashedSemaphore class was designed for the Observer class and it has little reuse value outside of it.

All in all, I think that this particular case is a good fit for a nested class.

**Comment by noxdafox** [comment]
Issue is the `@handler` decorator does more than that. First question which would come to my mind is: "what about non decorated handlers? Are they different? How?". 

What if, instead of using a decorator, we add a new subscriber method?
 - `subscribe`: Subscribes a Handler for the given Event.
 - `subscribe_async`: Subscribes an asynchronous Handler for the given Event.
 - `subscribe_refcount/mutex`: Subscribe a reference counted Handler for the given Event. A reference counted Handler is a handler which....

I like the idea of decorators but I'd prefer them to be applied to all handler and not to special ones only. A SEE 2.0 could re-factor the interface as:
 - `@handler("pre_poweron")`: A Handler for a `pre_poweron` event
 - `@async_handler("pre_poweron", "post_poweroff")`: An asynchronous Handler for 2 events
 - `@refcount_handler("poweroff")`: A reference counted Handler ...

Issue with the decoration method is it does not allow to make events configurable. Decorators accept only static variables.

**Comment by Jorge-Rodriguez** [comment]
The use case that prompted the creation of this PR can be serviced in a different manner with the currently existing functionality. In the name of keeping SEE lean, I will close this PR.

**Pull request closed without merging**